In [1]:
from platform import python_version

In [2]:
!pip install -q pandas == 1.5.3

ERROR: Invalid requirement: '=='


In [3]:
import pandas as pd

In [4]:
pd.__version__

'1.3.5'

In [5]:
#Criando um dicionário
dados = {'Estados':['Santa Catarina','Rio de Janeiro','Tocantins','Bahia','Minas Gerais'],
        'Ano':['2004','2005','2006','2007','2008'],
        'Taxa Desemprego':[1.5,1.7,1.6,2.4,2.7]}

In [6]:
#Importando a funçãio DataFrame no pandas
from pandas import DataFrame

In [7]:
#Convertendo um Dicionário em Data Frame
df = DataFrame(dados)

In [8]:
df

Estados   Ano  Taxa Desemprego
0  Santa Catarina  2004              1.5
1  Rio de Janeiro  2005              1.7
2       Tocantins  2006              1.6
3           Bahia  2007              2.4
4    Minas Gerais  2008              2.7

In [9]:
df.head()

Estados   Ano  Taxa Desemprego
0  Santa Catarina  2004              1.5
1  Rio de Janeiro  2005              1.7
2       Tocantins  2006              1.6
3           Bahia  2007              2.4
4    Minas Gerais  2008              2.7

In [10]:
type(df)

pandas.core.frame.DataFrame

In [11]:
#Reorganizando as colunas
DataFrame(dados,columns = ['Estados','Taxa Desemprego','Ano'])

Estados  Taxa Desemprego   Ano
0  Santa Catarina              1.5  2004
1  Rio de Janeiro              1.7  2005
2       Tocantins              1.6  2006
3           Bahia              2.4  2007
4    Minas Gerais              2.7  2008

In [12]:
#Criando um novo DataFrame com os dados anteriores porem inserindo mais uma coluna
df2 = DataFrame(dados, columns=['Estados','Taxa Desemprego','Taxa Crescimento','Ano'],
               index = ['estado1','estado2','estado3','estado4','estado5'])

In [13]:
df2

Estados  Taxa Desemprego Taxa Crescimento   Ano
estado1  Santa Catarina              1.5              NaN  2004
estado2  Rio de Janeiro              1.7              NaN  2005
estado3       Tocantins              1.6              NaN  2006
estado4           Bahia              2.4              NaN  2007
estado5    Minas Gerais              2.7              NaN  2008

In [14]:
df2.columns

Index(['Estados', 'Taxa Desemprego', 'Taxa Crescimento', 'Ano'], dtype='object')

In [15]:
df2.values

array([['Santa Catarina', 1.5, nan, '2004'],
       ['Rio de Janeiro', 1.7, nan, '2005'],
       ['Tocantins', 1.6, nan, '2006'],
       ['Bahia', 2.4, nan, '2007'],
       ['Minas Gerais', 2.7, nan, '2008']], dtype=object)

In [16]:
df2.dtypes

Estados              object
Taxa Desemprego     float64
Taxa Crescimento     object
Ano                  object
dtype: object

In [17]:
df2['Ano']=df2['Ano'].astype(int)

In [18]:
df2['Estados']

estado1    Santa Catarina
estado2    Rio de Janeiro
estado3         Tocantins
estado4             Bahia
estado5      Minas Gerais
Name: Estados, dtype: object

In [19]:
df2[['Ano','Estados']]

Ano         Estados
estado1  2004  Santa Catarina
estado2  2005  Rio de Janeiro
estado3  2006       Tocantins
estado4  2007           Bahia
estado5  2008    Minas Gerais

In [20]:
df2.index

Index(['estado1', 'estado2', 'estado3', 'estado4', 'estado5'], dtype='object')

In [21]:
df2.filter(items=['estado3'], axis=0)

Estados  Taxa Desemprego Taxa Crescimento   Ano
estado3  Tocantins              1.6              NaN  2006

In [22]:
#Resumo estatístico do DataFrame
df2.describe()

Taxa Desemprego          Ano
count         5.000000     5.000000
mean          1.980000  2006.000000
std           0.535724     1.581139
min           1.500000  2004.000000
25%           1.600000  2005.000000
50%           1.700000  2006.000000
75%           2.400000  2007.000000
max           2.700000  2008.000000

In [23]:
df2.isna()

Estados  Taxa Desemprego  Taxa Crescimento    Ano
estado1    False            False              True  False
estado2    False            False              True  False
estado3    False            False              True  False
estado4    False            False              True  False
estado5    False            False              True  False

In [24]:
df2['Taxa Crescimento'].isna()

estado1    True
estado2    True
estado3    True
estado4    True
estado5    True
Name: Taxa Crescimento, dtype: bool

In [25]:
import numpy as np

In [26]:
#Usando o numpy para preencher uma das colunas com valores ausent
df2['Taxa Crescimento'] = np.arange(5,)

In [27]:
df2['Taxa Crescimento'].isna()

estado1    False
estado2    False
estado3    False
estado4    False
estado5    False
Name: Taxa Crescimento, dtype: bool

In [28]:
df2

Estados  Taxa Desemprego  Taxa Crescimento   Ano
estado1  Santa Catarina              1.5                 0  2004
estado2  Rio de Janeiro              1.7                 1  2005
estado3       Tocantins              1.6                 2  2006
estado4           Bahia              2.4                 3  2007
estado5    Minas Gerais              2.7                 4  2008

In [29]:
df2.describe()

Taxa Desemprego  Taxa Crescimento          Ano
count         5.000000          5.000000     5.000000
mean          1.980000          2.000000  2006.000000
std           0.535724          1.581139     1.581139
min           1.500000          0.000000  2004.000000
25%           1.600000          1.000000  2005.000000
50%           1.700000          2.000000  2006.000000
75%           2.400000          3.000000  2007.000000
max           2.700000          4.000000  2008.000000

In [30]:
df2['estado2':'estado4']

Estados  Taxa Desemprego  Taxa Crescimento   Ano
estado2  Rio de Janeiro              1.7                 1  2005
estado3       Tocantins              1.6                 2  2006
estado4           Bahia              2.4                 3  2007

In [31]:
df2[df2['Taxa Desemprego']<2]

Estados  Taxa Desemprego  Taxa Crescimento   Ano
estado1  Santa Catarina              1.5                 0  2004
estado2  Rio de Janeiro              1.7                 1  2005
estado3       Tocantins              1.6                 2  2006

In [32]:
df2[['Taxa Crescimento','Estados']]

Taxa Crescimento         Estados
estado1                 0  Santa Catarina
estado2                 1  Rio de Janeiro
estado3                 2       Tocantins
estado4                 3           Bahia
estado5                 4    Minas Gerais

## Preenchendo valores ausentes com técnica de interpolação

In [24]:
#Importando o Dataset
dsa_df = pd.read_csv("dataset.csv")

In [25]:
dsa_df

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0     CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1     CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2     CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3     US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4     US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
...              ...         ...        ...        ...            ...    ...   
9989  CA-2014-110422  2014-01-21   TB-21400   Consumer  United States  South   
9990  CA-2017-121258  2017-02-26   DB-13060   Consumer  United States   West   
9991  CA-2017-121258  2017-02-26   DB-13060   Consumer  United States   West   
9992  CA-2017-121258  2017-02-26   DB-13060   Consumer  United States   West   
9993  CA-2017-119914  2017-05-04   CC-12220   Consumer  United States   West   

           ID_Produto        Categoria  \
0     FUR-BO-10001798        Furniture   
1     FUR-CH-10000454        Furniture   
2     OFF-LA-10000240  Office Supplies   
3     FUR-TA-10000577        Furniture   
4     OFF-ST-10000760  Office Supplies   
...               ...              ...   
9989  FUR-FU-10001889        Furniture   
9990  FUR-FU-10000747        Furniture   
9991  TEC-PH-10003645       Technology   
9992  OFF-PA-10004041  Office Supplies   
9993  OFF-AP-10002684  Office Supplies   

                                           Nome_Produto  Valor_Venda  \
0                     Bush Somerset Collection Bookcase     261.9600   
1     Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400   
2     Self-Adhesive Address Labels for Typewriters b...      14.6200   
3         Bretford CR4500 Series Slim Rectangular Table     957.5775   
4                        Eldon Fold 'N Roll Cart System      22.3680   
...                                                 ...          ...   
9989                             Ultra Door Pull Handle      25.2480   
9990  Tenex B1-RE Series Chair Mats for Low Pile Car...      91.9600   
9991                              Aastra 57i VoIP phone     258.5760   
9992  It's Hot Message Books with Stickers, 2 3/4" x 5"      29.6000   
9993  Acco 7-Outlet Masterpiece Power Center, Wihtou...     243.1600   

      Quantidade  
0            NaN  
1            NaN  
2            2.0  
3            5.0  
4            2.0  
...          ...  
9989         3.0  
9990         2.0  
9991         2.0  
9992         4.0  
9993         2.0  

[9994 rows x 11 columns]

In [26]:
    dsa_df.head(5)

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         NaN  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         NaN  
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0

In [27]:
dsa_df.isnull().sum()

ID_Pedido       0
Data_Pedido     0
ID_Cliente      0
Segmento        0
Pais            0
Regiao          0
ID_Produto      0
Categoria       0
Nome_Produto    0
Valor_Venda     0
Quantidade      2
dtype: int64

In [28]:
#Extraímos a moda da coluna Quantidade
moda = dsa_df['Quantidade'].value_counts().index[0]

## A moda estatistica é uma medida de tendencia central que representa o valor mais frequente em um conjunto de dados. A moda é especialmente útil quando queremos saber qual é o valor mais comum ou popular em um conjunto de dados, seja em uma distribuição unimodal (com apenas uma medida) ou uma distrobuição bimodal (com duas modas). No entando a moda pode não ser uma medida tão representativa quanto outras medidas de tendencia central como a média ou a mediana, especialmente em distribuições assimétricas ou quando a valores extremos. Por essa razão é importante analisar diferentes medidas de tendencia central e usar a que melhor se adequa aos seus objetivos na análise estatistica.

In [29]:
print(moda)

3.0


In [30]:
#E por fim preenchemos os valores na com a moda
dsa_df['Quantidade'].fillna(value = moda, inplace = True)

In [31]:
dsa_df.isna().sum()

ID_Pedido       0
Data_Pedido     0
ID_Cliente      0
Segmento        0
Pais            0
Regiao          0
ID_Produto      0
Categoria       0
Nome_Produto    0
Valor_Venda     0
Quantidade      0
dtype: int64

### Query - Consulta no Data Frame do Pandas
##### Com o Pandas criamos Data Frames, que são essencialmente tabelas. Como tal podemos fazer consultas, ou simplesmente queries. E para isso utilizamos o método query(). Veja o exemplo abaixo:

In [32]:
#Checamos o valor minimo e máximo de uma coluna valor vendas
dsa_df.Valor_Venda.describe()

count     9994.000000
mean       229.858001
std        623.245101
min          0.444000
25%         17.280000
50%         54.490000
75%        209.940000
max      22638.480000
Name: Valor_Venda, dtype: float64

###### O intervalo de vendas (Valor_venda)  é de 0.44 a 22638. Vamos fazer  uma consulta e retornar todas as vendas dentro de um range de valores. Fazemos isso com a instrução abaixo:

In [33]:
dsa_df2 = dsa_df.query('229 < Valor_Venda < 10000')

In [34]:
dsa_df2

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
0     CA-2016-152156  2016-11-08   CG-12520     Consumer  United States   
1     CA-2016-152156  2016-11-08   CG-12520     Consumer  United States   
3     US-2015-108966  2015-10-11   SO-20335     Consumer  United States   
7     CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
10    CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
...              ...         ...        ...          ...            ...   
9973  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9976  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9979  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9991  CA-2017-121258  2017-02-26   DB-13060     Consumer  United States   
9993  CA-2017-119914  2017-05-04   CC-12220     Consumer  United States   

     Regiao       ID_Produto        Categoria  \
0     South  FUR-BO-10001798        Furniture   
1     South  FUR-CH-10000454        Furniture   
3     South  FUR-TA-10000577        Furniture   
7      West  TEC-PH-10002275       Technology   
10     West  FUR-TA-10001539        Furniture   
...     ...              ...              ...   
9973   West  TEC-PH-10004080       Technology   
9976   West  TEC-PH-10002496       Technology   
9979   West  OFF-BI-10002026  Office Supplies   
9991   West  TEC-PH-10003645       Technology   
9993   West  OFF-AP-10002684  Office Supplies   

                                           Nome_Produto  Valor_Venda  \
0                     Bush Somerset Collection Bookcase     261.9600   
1     Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400   
3         Bretford CR4500 Series Slim Rectangular Table     957.5775   
7                        Mitel 5320 IP Phone VoIP phone     907.1520   
10             Chromcraft Rectangular Conference Tables    1706.1840   
...                                                 ...          ...   
9973                           Avaya 5410 Digital phone     271.9600   
9976                                       Cisco SPA301     249.5840   
9979                  Ibico Recycled Linen-Style Covers     437.4720   
9991                              Aastra 57i VoIP phone     258.5760   
9993  Acco 7-Outlet Masterpiece Power Center, Wihtou...     243.1600   

      Quantidade  
0            3.0  
1            3.0  
3            5.0  
7            6.0  
10           9.0  
...          ...  
9973         5.0  
9976         2.0  
9979        14.0  
9991         2.0  
9993         2.0  

[2357 rows x 11 columns]

In [35]:
dsa_df2.describe()

Valor_Venda   Quantidade
count  2357.000000  2357.000000
mean    766.679142     4.654221
std     856.315136     2.438478
min     229.544000     1.000000
25%     323.100000     3.000000
50%     490.320000     4.000000
75%     859.200000     6.000000
max    9892.740000    14.000000

In [36]:
dsa_df3 = dsa_df2.query('Valor_Venda > 766')

In [37]:
dsa_df3.describe()

Valor_Venda  Quantidade
count   687.000000  687.000000
mean   1615.597817    5.599709
std    1203.403998    2.498081
min     767.214000    1.000000
25%     914.200000    4.000000
50%    1212.960000    5.000000
75%    1763.575000    7.000000
max    9892.740000   14.000000

In [38]:
dsa_df3

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais  \
3     US-2015-108966  2015-10-11   SO-20335   Consumer  United States   
7     CA-2014-115812  2014-06-09   BH-11710   Consumer  United States   
10    CA-2014-115812  2014-06-09   BH-11710   Consumer  United States   
11    CA-2014-115812  2014-06-09   BH-11710   Consumer  United States   
24    CA-2015-106320  2015-09-25   EB-13870   Consumer  United States   
...              ...         ...        ...        ...            ...   
9925  CA-2015-159534  2015-03-20   DH-13075  Corporate  United States   
9929  CA-2016-129630  2016-09-04   IM-15055   Consumer  United States   
9942  CA-2014-143371  2014-12-28   MD-17350   Consumer  United States   
9947  CA-2017-121559  2017-06-01   HW-14935  Corporate  United States   
9948  CA-2017-121559  2017-06-01   HW-14935  Corporate  United States   

       Regiao       ID_Produto        Categoria  \
3       South  FUR-TA-10000577        Furniture   
7        West  TEC-PH-10002275       Technology   
10       West  FUR-TA-10001539        Furniture   
11       West  TEC-PH-10002033       Technology   
24       West  FUR-TA-10000577        Furniture   
...       ...              ...              ...   
9925     East  OFF-BI-10003656  Office Supplies   
9929     West  TEC-CO-10003763       Technology   
9942     West  OFF-ST-10001128  Office Supplies   
9947  Central  FUR-CH-10003746        Furniture   
9948  Central  OFF-AP-10002945  Office Supplies   

                                           Nome_Produto  Valor_Venda  \
3         Bretford CR4500 Series Slim Rectangular Table     957.5775   
7                        Mitel 5320 IP Phone VoIP phone     907.1520   
10             Chromcraft Rectangular Conference Tables    1706.1840   
11        Konftel 250 Conference phone - Charcoal black     911.4240   
24        Bretford CR4500 Series Slim Rectangular Table    1044.6300   
...                                                 ...          ...   
9925        Fellowes PB200 Plastic Comb Binding Machine    1087.9360   
9929                 Canon PC1060 Personal Laser Copier    2799.9600   
9942        Carina Mini System Audio Rack, Model AR050B     998.8200   
9947  Hon 4070 Series Pagoda Round Back Stacking Chairs    1925.8800   
9948  Honeywell Enviracaire Portable HEPA Air Cleane...    2405.2000   

      Quantidade  
3            5.0  
7            6.0  
10           9.0  
11           4.0  
24           3.0  
...          ...  
9925         8.0  
9929         5.0  
9942         9.0  
9947         6.0  
9948         8.0  

[687 rows x 11 columns]

# Verificando a ocorrencia de diversos valores em uma coluna.
##### Em nosso conjunto de dados temos a coluna [Quantidade] que representa a quantodade itens vendidos em cada uma das vendas.
##### Imagine que precisamos saber em quais vendas foram vendidos, 5 , 7 , 9 e 11 itens. Como aplicariamos esse tipo de filtro em nosso Data Frame?.
##### O pandas oferece o modo isin() para checar diversos valores em uma coluna. Quem conhece a linguagem SQL já deve ter percebido que o método é o mesmo da clausula IN no SQL.

In [39]:
dsa_df.shape

(9994, 11)

In [40]:
#Então aplicamos o filtro a coluna [Quantidade]
dsa_df[dsa_df['Quantidade'].isin([5,7,9,11])]

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
3     US-2015-108966  2015-10-11   SO-20335     Consumer  United States   
5     CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
9     CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
10    CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
14    US-2015-118983  2015-11-22   HP-14815  Home Office  United States   
...              ...         ...        ...          ...            ...   
9974  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9977  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9981  CA-2017-163566  2017-08-03   TB-21055     Consumer  United States   
9982  US-2016-157728  2016-09-22   RC-19960     Consumer  United States   
9988  CA-2017-163629  2017-11-17   RA-19885    Corporate  United States   

       Regiao       ID_Produto        Categoria  \
3       South  FUR-TA-10000577        Furniture   
5        West  FUR-FU-10001487        Furniture   
9        West  OFF-AP-10002892  Office Supplies   
10       West  FUR-TA-10001539        Furniture   
14    Central  OFF-AP-10002311  Office Supplies   
...       ...              ...              ...   
9974     West  OFF-AR-10004752  Office Supplies   
9977     West  OFF-FA-10003467  Office Supplies   
9981     East  OFF-LA-10004484  Office Supplies   
9982  Central  OFF-PA-10002195  Office Supplies   
9988    South  TEC-PH-10004006       Technology   

                                           Nome_Produto  Valor_Venda  \
3         Bretford CR4500 Series Slim Rectangular Table     957.5775   
5     Eldon Expressions Wood and Plastic Desk Access...      48.8600   
9                      Belkin F5C206VTEL 6 Outlet Surge     114.9000   
10             Chromcraft Rectangular Conference Tables    1706.1840   
14    Holmes Replacement Filter for HEPA Air Cleaner...      68.8100   
...                                                 ...          ...   
9974                               Blackstonian Pencils      18.6900   
9977           Alliance Big Bands Rubber Bands, 12/Pack      13.8600   
9981                                          Avery 476      16.5200   
9982  RSVP Cards & Envelopes, Blank White, 8-1/2" X ...      35.5600   
9988                    Panasonic KX - TS880B Telephone     206.1000   

      Quantidade  
3            5.0  
5            7.0  
9            5.0  
10           9.0  
14           5.0  
...          ...  
9974         7.0  
9977         7.0  
9981         5.0  
9982         7.0  
9988         5.0  

[2128 rows x 11 columns]

In [65]:
dsa_2ff = dsa_df.query('Segmento == "Consumer"')

###### Na instrução acima estamos filtrando o DataFrame chamado dsa_df, retornando todas as linhas onde a coluna quantidade tem os valores (5,7,9,11). Passamos uma lista de valores como argumento para o método isin().
##### Vamos deixar um pouquinho mais divertido. E se quisermos retornar apenas 10 linhas?

In [73]:
dsa_3ff = dsa_df[dsa_df['Segmento'] == "Consumer"]

In [75]:
dsa_4ff = dsa_df[dsa_df['Categoria'] == "Technology"]

In [76]:
dsa_4ff

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais  \
7     CA-2014-115812  2014-06-09   BH-11710   Consumer  United States   
11    CA-2014-115812  2014-06-09   BH-11710   Consumer  United States   
19    CA-2014-143336  2014-08-27   ZD-21925   Consumer  United States   
26    CA-2016-121755  2016-01-16   EH-13945   Consumer  United States   
35    CA-2016-117590  2016-12-08   GH-14485  Corporate  United States   
...              ...         ...        ...        ...            ...   
9983  US-2016-157728  2016-09-22   RC-19960   Consumer  United States   
9986  CA-2016-125794  2016-09-29   ML-17410   Consumer  United States   
9987  CA-2017-163629  2017-11-17   RA-19885  Corporate  United States   
9988  CA-2017-163629  2017-11-17   RA-19885  Corporate  United States   
9991  CA-2017-121258  2017-02-26   DB-13060   Consumer  United States   

       Regiao       ID_Produto   Categoria  \
7        West  TEC-PH-10002275  Technology   
11       West  TEC-PH-10002033  Technology   
19       West  TEC-PH-10001949  Technology   
26       West  TEC-AC-10003027  Technology   
35    Central  TEC-PH-10004977  Technology   
...       ...              ...         ...   
9983  Central  TEC-PH-10001305  Technology   
9986     West  TEC-AC-10003399  Technology   
9987    South  TEC-AC-10001539  Technology   
9988    South  TEC-PH-10004006  Technology   
9991     West  TEC-PH-10003645  Technology   

                                           Nome_Produto  Valor_Venda  \
7                        Mitel 5320 IP Phone VoIP phone      907.152   
11        Konftel 250 Conference phone - Charcoal black      911.424   
19                              Cisco SPA 501G IP Phone      213.480   
26     Imation 8GB Mini TravelDrive USB 2.0 Flash Drive       90.570   
35                                          GE 30524EE4     1097.544   
...                                                 ...          ...   
9983                   Panasonic KX TS208W Corded phone       97.980   
9986  Memorex Mini Travel Drive 64 GB USB 2.0 Flash ...       36.240   
9987  Logitech G430 Surround Sound Gaming Headset wi...       79.990   
9988                    Panasonic KX - TS880B Telephone      206.100   
9991                              Aastra 57i VoIP phone      258.576   

      Quantidade  
7            6.0  
11           4.0  
19           3.0  
26           3.0  
35           7.0  
...          ...  
9983         2.0  
9986         1.0  
9987         1.0  
9988         5.0  
9991         2.0  

[1847 rows x 11 columns]

In [50]:
#aplicando o mesmo filtro, porem retornando as 10 primeiras linhas
dsa_df[dsa_df['Quantidade'].isin([5,7,9,11])][:10]

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
3   US-2015-108966  2015-10-11   SO-20335     Consumer  United States   
5   CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
9   CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
10  CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
14  US-2015-118983  2015-11-22   HP-14815  Home Office  United States   
21  CA-2016-137330  2016-12-09   KB-16585    Corporate  United States   
22  CA-2016-137330  2016-12-09   KB-16585    Corporate  United States   
27  US-2015-150630  2015-09-17   TB-21520     Consumer  United States   
35  CA-2016-117590  2016-12-08   GH-14485    Corporate  United States   
36  CA-2016-117590  2016-12-08   GH-14485    Corporate  United States   

     Regiao       ID_Produto        Categoria  \
3     South  FUR-TA-10000577        Furniture   
5      West  FUR-FU-10001487        Furniture   
9      West  OFF-AP-10002892  Office Supplies   
10     West  FUR-TA-10001539        Furniture   
14  Central  OFF-AP-10002311  Office Supplies   
21  Central  OFF-AR-10000246  Office Supplies   
22  Central  OFF-AP-10001492  Office Supplies   
27     East  FUR-BO-10004834        Furniture   
35  Central  TEC-PH-10004977       Technology   
36  Central  FUR-FU-10003664        Furniture   

                                         Nome_Produto  Valor_Venda  Quantidade  
3       Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
5   Eldon Expressions Wood and Plastic Desk Access...      48.8600         7.0  
9                    Belkin F5C206VTEL 6 Outlet Surge     114.9000         5.0  
10           Chromcraft Rectangular Conference Tables    1706.1840         9.0  
14  Holmes Replacement Filter for HEPA Air Cleaner...      68.8100         5.0  
21                                         Newell 318      19.4600         7.0  
22        Acco Six-Outlet Power Strip, 4' Cord Length      60.3400         7.0  
27  Riverside Palais Royal Lawyers Bookcase, Royal...    3083.4300         7.0  
35                                        GE 30524EE4    1097.5440         7.0  
36  Electrix Architect's Clamp-On Swing Arm Lamp, ...     190.9200         5.0

In [51]:
#shape
dsa_df[dsa_df['Quantidade'].isin([5,7,9,11])].shape

(2128, 11)

## Operadores lógicos para manipuação de dados com o Pandas.


##### Operador AND

In [52]:
#Filtrando as vendas que ocorrem no segmento home-office na região sul
# O [] indica o slice e o () é onde colocamos a expressão, essa é a regra para aplcarmos expressões dentro de slices, usando 
# o dataframe.coluna dentro de parenteses e todo o slice dentro de colchetes.
#O & representa o operador AND
dsa_df4 = dsa_df[(dsa_df.Segmento == 'Home Office') & (dsa_df.Regiao=='South')].head()

In [53]:
dsa_df4

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais Regiao  \
182  CA-2014-158274  2014-11-19   RM-19675  Home Office  United States  South   
183  CA-2014-158274  2014-11-19   RM-19675  Home Office  United States  South   
184  CA-2014-158274  2014-11-19   RM-19675  Home Office  United States  South   
231  US-2017-100930  2017-04-07   CS-12400  Home Office  United States  South   
232  US-2017-100930  2017-04-07   CS-12400  Home Office  United States  South   

          ID_Produto   Categoria  \
182  TEC-PH-10003273  Technology   
183  TEC-PH-10004896  Technology   
184  TEC-AC-10002345  Technology   
231  FUR-TA-10001705   Furniture   
232  FUR-TA-10003473   Furniture   

                                         Nome_Produto  Valor_Venda  Quantidade  
182                                      AT&T TR1909W     503.9600         4.0  
183                        Nokia Lumia 521 (T-Mobile)     149.9500         5.0  
184                 HP Standard 104 key PS/2 Keyboard      29.0000         2.0  
231  Bush Advantage Collection Round Conference Table     233.8600         2.0  
232        Bretford Rectangular Conference Table Tops     620.6145         3.0

##### Operador OR

In [54]:
#Filtrando as vendas que ocorrem no segmento home-office ou na região sul
# O [] indica o slice e o () é onde colocamos a expressão, essa é a regra para aplcarmos expressões dentro de slices, usando 
# o dataframe.coluna dentro de parenteses e todo o slice dentro de colchetes.
#O | representa o operador OR
dsa_df5 = dsa_df[(dsa_df.Segmento == 'Home Office') | (dsa_df.Regiao=='South')].tail()

In [55]:
dsa_df5

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
9979  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9980  US-2015-151435  2015-09-06   SW-20455     Consumer  United States   
9987  CA-2017-163629  2017-11-17   RA-19885    Corporate  United States   
9988  CA-2017-163629  2017-11-17   RA-19885    Corporate  United States   
9989  CA-2014-110422  2014-01-21   TB-21400     Consumer  United States   

     Regiao       ID_Produto        Categoria  \
9979   West  OFF-BI-10002026  Office Supplies   
9980  South  FUR-TA-10001029        Furniture   
9987  South  TEC-AC-10001539       Technology   
9988  South  TEC-PH-10004006       Technology   
9989  South  FUR-FU-10001889        Furniture   

                                           Nome_Produto  Valor_Venda  \
9979                  Ibico Recycled Linen-Style Covers      437.472   
9980                         KI Adjustable-Height Table       85.980   
9987  Logitech G430 Surround Sound Gaming Headset wi...       79.990   
9988                    Panasonic KX - TS880B Telephone      206.100   
9989                             Ultra Door Pull Handle       25.248   

      Quantidade  
9979        14.0  
9980         1.0  
9987         1.0  
9988         5.0  
9989         3.0

##### Operador Negação

In [56]:
# != significa "Diferente em Python"
#Método Sample retorna amostras aleatórias
dsa_df6 = dsa_df[(dsa_df.Categoria != 'Home Office') & (dsa_df.Regiao != 'South')].sample(5)

In [57]:
dsa_df6

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
2877  CA-2016-152072  2016-01-15   Dp-13240  Home Office  United States   
161   CA-2015-119697  2015-12-28   EM-13960     Consumer  United States   
3880  CA-2015-129896  2015-06-15   PF-19120     Consumer  United States   
8961  CA-2017-150266  2017-11-25   RO-19780     Consumer  United States   
201   CA-2014-133690  2014-08-03   BS-11755     Consumer  United States   

       Regiao       ID_Produto        Categoria  \
2877     East  OFF-EN-10003040  Office Supplies   
161      East  TEC-AC-10003657       Technology   
3880     West  OFF-PA-10002245  Office Supplies   
8961  Central  OFF-AR-10001761  Office Supplies   
201      West  FUR-TA-10004289        Furniture   

                                           Nome_Produto  Valor_Venda  \
2877                    Quality Park Security Envelopes       52.340   
161                    Lenovo 17-Key USB Numeric Keypad       54.384   
3880                                         Xerox 1895        9.568   
8961             Avery Hi-Liter Smear-Safe Highlighters       18.688   
201   BoxOffice By Design Rectangular and Half-Moon ...      218.750   

      Quantidade  
2877         2.0  
161          2.0  
3880         2.0  
8961         4.0  
201          2.0

### Agrupamento de dados em DataFrames com o GroupBy

In [58]:
#Aplicando o grupBy - Seria o mesmo que fazer uma tabela Dinamuica no Excel 
#O agrupamento precisa ser por colunas não numéricas
#O groupBy precisa de uma função matemática de uma coluna que não aparece no agrupamento, ou seja, uma coluna numerica.
dsa_df[['Segmento','Regiao','Valor_Venda']].groupby(['Segmento','Regiao']).mean()

Valor_Venda
Segmento    Regiao              
Consumer    Central   207.946728
            East      238.875539
            South     233.390180
            West      217.033955
Corporate   Central   234.763466
            East      228.516929
            South     238.992025
            West      235.265911
Home Office Central   208.248046
            East      253.911805
            South     272.996329
            West      239.442692

In [59]:
dsa_df[['Regiao','Categoria','Quantidade']].groupby(['Regiao','Categoria']).count()

Quantidade
Regiao  Categoria                  
Central Furniture               481
        Office Supplies        1422
        Technology              420
East    Furniture               601
        Office Supplies        1712
        Technology              535
South   Furniture               332
        Office Supplies         995
        Technology              293
West    Furniture               707
        Office Supplies        1897
        Technology              599

In [60]:
dsa_df.dtypes

ID_Pedido        object
Data_Pedido      object
ID_Cliente       object
Segmento         object
Pais             object
Regiao           object
ID_Produto       object
Categoria        object
Nome_Produto     object
Valor_Venda     float64
Quantidade      float64
dtype: object

#### Agragação multipla com o GroupBY - Vamos explorar mais a função groupby() pois temos diversas opções de sumarização
#### de dados de forma simples. No exemplo abaixo uniremos a função groupBy com a função agg que realiza agragação multipla.

In [61]:
#Aplicanod groupby + agg
dsa_df[['Segmento','Regiao','Valor_Venda']].groupby(['Segmento','Regiao']).agg(['mean','std','count'])

Valor_Venda                   
                           mean          std count
Segmento    Regiao                                
Consumer    Central  207.946728   587.906523  1212
            East     238.875539   633.371169  1469
            South    233.390180   559.346824   838
            West     217.033955   551.997547  1672
Corporate   Central  234.763466   818.947521   673
            East     228.516929   530.001654   877
            South    238.992025   586.176947   510
            West     235.265911   471.288764   960
Home Office Central  208.248046   371.009180   438
            East     253.911805   722.777318   502
            South    272.996329  1404.798466   272
            West     239.442692   529.242737   571

In [62]:
#Filtrando a partir d strings - Utilizando o método startswith -> filtra pelo começo de uma string, ex: startswith 'Con'
dsa_df[dsa_df.Segmento.str.startswith('Con')]

ID_Pedido Data_Pedido ID_Cliente  Segmento           Pais Regiao  \
0     CA-2016-152156  2016-11-08   CG-12520  Consumer  United States  South   
1     CA-2016-152156  2016-11-08   CG-12520  Consumer  United States  South   
3     US-2015-108966  2015-10-11   SO-20335  Consumer  United States  South   
4     US-2015-108966  2015-10-11   SO-20335  Consumer  United States  South   
5     CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   
...              ...         ...        ...       ...            ...    ...   
9989  CA-2014-110422  2014-01-21   TB-21400  Consumer  United States  South   
9990  CA-2017-121258  2017-02-26   DB-13060  Consumer  United States   West   
9991  CA-2017-121258  2017-02-26   DB-13060  Consumer  United States   West   
9992  CA-2017-121258  2017-02-26   DB-13060  Consumer  United States   West   
9993  CA-2017-119914  2017-05-04   CC-12220  Consumer  United States   West   

           ID_Produto        Categoria  \
0     FUR-BO-10001798        Furniture   
1     FUR-CH-10000454        Furniture   
3     FUR-TA-10000577        Furniture   
4     OFF-ST-10000760  Office Supplies   
5     FUR-FU-10001487        Furniture   
...               ...              ...   
9989  FUR-FU-10001889        Furniture   
9990  FUR-FU-10000747        Furniture   
9991  TEC-PH-10003645       Technology   
9992  OFF-PA-10004041  Office Supplies   
9993  OFF-AP-10002684  Office Supplies   

                                           Nome_Produto  Valor_Venda  \
0                     Bush Somerset Collection Bookcase     261.9600   
1     Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400   
3         Bretford CR4500 Series Slim Rectangular Table     957.5775   
4                        Eldon Fold 'N Roll Cart System      22.3680   
5     Eldon Expressions Wood and Plastic Desk Access...      48.8600   
...                                                 ...          ...   
9989                             Ultra Door Pull Handle      25.2480   
9990  Tenex B1-RE Series Chair Mats for Low Pile Car...      91.9600   
9991                              Aastra 57i VoIP phone     258.5760   
9992  It's Hot Message Books with Stickers, 2 3/4" x 5"      29.6000   
9993  Acco 7-Outlet Masterpiece Power Center, Wihtou...     243.1600   

      Quantidade  
0            3.0  
1            3.0  
3            5.0  
4            2.0  
5            7.0  
...          ...  
9989         3.0  
9990         2.0  
9991         2.0  
9992         4.0  
9993         2.0  

[5191 rows x 11 columns]

In [63]:
dsa_df.Segmento.value_counts()

Consumer       5191
Corporate      3020
Home Office    1783
Name: Segmento, dtype: int64

In [64]:
#Filtrando a partir d strings - Utilizando o método endswith -> filtra pelo final de uma string, ex: endswith 'mer'
dsa_df[dsa_df.Segmento.str.endswith('mer')].head()

ID_Pedido Data_Pedido ID_Cliente  Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520  Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520  Consumer  United States  South   
3  US-2015-108966  2015-10-11   SO-20335  Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335  Consumer  United States  South   
5  CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   
5  FUR-FU-10001487        Furniture   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         3.0  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0  
5  Eldon Expressions Wood and Plastic Desk Access...      48.8600         7.0

## Split de Strings em DataFrames Pandas.
Com o Pandas podemos realizar diversas tarfas de split de string dividindo uma coluna ou extraindo elementos do nosso interesse.

In [65]:
dsa_df.head()

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         3.0  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0  
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0

In [66]:
dsa_df['ID_Pedido'].head()

0    CA-2016-152156
1    CA-2016-152156
2    CA-2016-138688
3    US-2015-108966
4    US-2015-108966
Name: ID_Pedido, dtype: object

In [67]:
#Split da coluna por caracter
#P resultado saõ listas de valores
dsa_df['ID_Pedido'].str.split("-")

0       [CA, 2016, 152156]
1       [CA, 2016, 152156]
2       [CA, 2016, 138688]
3       [US, 2015, 108966]
4       [US, 2015, 108966]
               ...        
9989    [CA, 2014, 110422]
9990    [CA, 2017, 121258]
9991    [CA, 2017, 121258]
9992    [CA, 2017, 121258]
9993    [CA, 2017, 119914]
Name: ID_Pedido, Length: 9994, dtype: object

In [68]:
#Pra extrair apenas o ano da lista de valores, usamos o fatiamento no Pandas
dsa_df['ID_Pedido'].str.split("-").str[1].head()

0    2016
1    2016
2    2016
3    2015
4    2015
Name: ID_Pedido, dtype: object

In [69]:
dsa_df['Ano']=dsa_df['ID_Pedido'].str.split("-").str[1]

In [70]:
dsa_df

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0     CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1     CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2     CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3     US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4     US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
...              ...         ...        ...        ...            ...    ...   
9989  CA-2014-110422  2014-01-21   TB-21400   Consumer  United States  South   
9990  CA-2017-121258  2017-02-26   DB-13060   Consumer  United States   West   
9991  CA-2017-121258  2017-02-26   DB-13060   Consumer  United States   West   
9992  CA-2017-121258  2017-02-26   DB-13060   Consumer  United States   West   
9993  CA-2017-119914  2017-05-04   CC-12220   Consumer  United States   West   

           ID_Produto        Categoria  \
0     FUR-BO-10001798        Furniture   
1     FUR-CH-10000454        Furniture   
2     OFF-LA-10000240  Office Supplies   
3     FUR-TA-10000577        Furniture   
4     OFF-ST-10000760  Office Supplies   
...               ...              ...   
9989  FUR-FU-10001889        Furniture   
9990  FUR-FU-10000747        Furniture   
9991  TEC-PH-10003645       Technology   
9992  OFF-PA-10004041  Office Supplies   
9993  OFF-AP-10002684  Office Supplies   

                                           Nome_Produto  Valor_Venda  \
0                     Bush Somerset Collection Bookcase     261.9600   
1     Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400   
2     Self-Adhesive Address Labels for Typewriters b...      14.6200   
3         Bretford CR4500 Series Slim Rectangular Table     957.5775   
4                        Eldon Fold 'N Roll Cart System      22.3680   
...                                                 ...          ...   
9989                             Ultra Door Pull Handle      25.2480   
9990  Tenex B1-RE Series Chair Mats for Low Pile Car...      91.9600   
9991                              Aastra 57i VoIP phone     258.5760   
9992  It's Hot Message Books with Stickers, 2 3/4" x 5"      29.6000   
9993  Acco 7-Outlet Masterpiece Power Center, Wihtou...     243.1600   

      Quantidade   Ano  
0            3.0  2016  
1            3.0  2016  
2            2.0  2016  
3            5.0  2015  
4            2.0  2015  
...          ...   ...  
9989         3.0  2014  
9990         2.0  2017  
9991         2.0  2017  
9992         4.0  2017  
9993         2.0  2017  

[9994 rows x 12 columns]

In [71]:
dsa_df.dtypes

ID_Pedido        object
Data_Pedido      object
ID_Cliente       object
Segmento         object
Pais             object
Regiao           object
ID_Produto       object
Categoria        object
Nome_Produto     object
Valor_Venda     float64
Quantidade      float64
Ano              object
dtype: object

In [72]:
dsa_df['Ano']=dsa_df['Ano'].astype(int)

In [73]:
dsa_df.dtypes

ID_Pedido        object
Data_Pedido      object
ID_Cliente       object
Segmento         object
Pais             object
Regiao           object
ID_Produto       object
Categoria        object
Nome_Produto     object
Valor_Venda     float64
Quantidade      float64
Ano               int32
dtype: object

In [74]:
dsa_df['ID_Pedido']=dsa_df['ID_Pedido'].str.split("-").str[2]

In [75]:
dsa_df

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0       152156  2016-11-08   CG-12520   Consumer  United States  South   
1       152156  2016-11-08   CG-12520   Consumer  United States  South   
2       138688  2016-06-12   DV-13045  Corporate  United States   West   
3       108966  2015-10-11   SO-20335   Consumer  United States  South   
4       108966  2015-10-11   SO-20335   Consumer  United States  South   
...        ...         ...        ...        ...            ...    ...   
9989    110422  2014-01-21   TB-21400   Consumer  United States  South   
9990    121258  2017-02-26   DB-13060   Consumer  United States   West   
9991    121258  2017-02-26   DB-13060   Consumer  United States   West   
9992    121258  2017-02-26   DB-13060   Consumer  United States   West   
9993    119914  2017-05-04   CC-12220   Consumer  United States   West   

           ID_Produto        Categoria  \
0     FUR-BO-10001798        Furniture   
1     FUR-CH-10000454        Furniture   
2     OFF-LA-10000240  Office Supplies   
3     FUR-TA-10000577        Furniture   
4     OFF-ST-10000760  Office Supplies   
...               ...              ...   
9989  FUR-FU-10001889        Furniture   
9990  FUR-FU-10000747        Furniture   
9991  TEC-PH-10003645       Technology   
9992  OFF-PA-10004041  Office Supplies   
9993  OFF-AP-10002684  Office Supplies   

                                           Nome_Produto  Valor_Venda  \
0                     Bush Somerset Collection Bookcase     261.9600   
1     Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400   
2     Self-Adhesive Address Labels for Typewriters b...      14.6200   
3         Bretford CR4500 Series Slim Rectangular Table     957.5775   
4                        Eldon Fold 'N Roll Cart System      22.3680   
...                                                 ...          ...   
9989                             Ultra Door Pull Handle      25.2480   
9990  Tenex B1-RE Series Chair Mats for Low Pile Car...      91.9600   
9991                              Aastra 57i VoIP phone     258.5760   
9992  It's Hot Message Books with Stickers, 2 3/4" x 5"      29.6000   
9993  Acco 7-Outlet Masterpiece Power Center, Wihtou...     243.1600   

      Quantidade   Ano  
0            3.0  2016  
1            3.0  2016  
2            2.0  2016  
3            5.0  2015  
4            2.0  2015  
...          ...   ...  
9989         3.0  2014  
9990         2.0  2017  
9991         2.0  2017  
9992         4.0  2017  
9993         2.0  2017  

[9994 rows x 12 columns]

### Strip de Strings em DataFrames do Pandas.
O strip remove caracteres da string.

In [76]:
dsa_df.head()

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0    152156  2016-11-08   CG-12520   Consumer  United States  South   
1    152156  2016-11-08   CG-12520   Consumer  United States  South   
2    138688  2016-06-12   DV-13045  Corporate  United States   West   
3    108966  2015-10-11   SO-20335   Consumer  United States  South   
4    108966  2015-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  \
0                  Bush Somerset Collection Bookcase     261.9600         3.0   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0   
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0   
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0   
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0   

    Ano  
0  2016  
1  2016  
2  2016  
3  2015  
4  2015

In [79]:
dsa_df['Data_Pedido']=dsa_df['Data_Pedido'].str.lstrip('20')

In [80]:
dsa_df.head()

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0    152156    16-11-08   CG-12520   Consumer  United States  South   
1    152156    16-11-08   CG-12520   Consumer  United States  South   
2    138688    16-06-12   DV-13045  Corporate  United States   West   
3    108966    15-10-11   SO-20335   Consumer  United States  South   
4    108966    15-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  \
0                  Bush Somerset Collection Bookcase     261.9600         3.0   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0   
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0   
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0   
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0   

    Ano  
0  2016  
1  2016  
2  2016  
3  2015  
4  2015

## Replace de strings em DataFrames pandas.
Se for necessário substituir caracteres em uma string, o pandas oferece função para isso tambem.

In [81]:
#Substituir dois caracteres em uma string.
#substituiiremos os caracteres 'CG' por 'AX' na coluna ID cliente
dsa_df['ID_Cliente'] = dsa_df['ID_Cliente'].str.replace('CG','AX')

In [82]:
dsa_df

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0       152156    16-11-08   AX-12520   Consumer  United States  South   
1       152156    16-11-08   AX-12520   Consumer  United States  South   
2       138688    16-06-12   DV-13045  Corporate  United States   West   
3       108966    15-10-11   SO-20335   Consumer  United States  South   
4       108966    15-10-11   SO-20335   Consumer  United States  South   
...        ...         ...        ...        ...            ...    ...   
9989    110422    14-01-21   TB-21400   Consumer  United States  South   
9990    121258    17-02-26   DB-13060   Consumer  United States   West   
9991    121258    17-02-26   DB-13060   Consumer  United States   West   
9992    121258    17-02-26   DB-13060   Consumer  United States   West   
9993    119914    17-05-04   CC-12220   Consumer  United States   West   

           ID_Produto        Categoria  \
0     FUR-BO-10001798        Furniture   
1     FUR-CH-10000454        Furniture   
2     OFF-LA-10000240  Office Supplies   
3     FUR-TA-10000577        Furniture   
4     OFF-ST-10000760  Office Supplies   
...               ...              ...   
9989  FUR-FU-10001889        Furniture   
9990  FUR-FU-10000747        Furniture   
9991  TEC-PH-10003645       Technology   
9992  OFF-PA-10004041  Office Supplies   
9993  OFF-AP-10002684  Office Supplies   

                                           Nome_Produto  Valor_Venda  \
0                     Bush Somerset Collection Bookcase     261.9600   
1     Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400   
2     Self-Adhesive Address Labels for Typewriters b...      14.6200   
3         Bretford CR4500 Series Slim Rectangular Table     957.5775   
4                        Eldon Fold 'N Roll Cart System      22.3680   
...                                                 ...          ...   
9989                             Ultra Door Pull Handle      25.2480   
9990  Tenex B1-RE Series Chair Mats for Low Pile Car...      91.9600   
9991                              Aastra 57i VoIP phone     258.5760   
9992  It's Hot Message Books with Stickers, 2 3/4" x 5"      29.6000   
9993  Acco 7-Outlet Masterpiece Power Center, Wihtou...     243.1600   

      Quantidade   Ano  
0            3.0  2016  
1            3.0  2016  
2            2.0  2016  
3            5.0  2015  
4            2.0  2015  
...          ...   ...  
9989         3.0  2014  
9990         2.0  2017  
9991         2.0  2017  
9992         4.0  2017  
9993         2.0  2017  

[9994 rows x 12 columns]

In [85]:
#Texte exportação do aqrquivo tratado
dsa_df.to_csv('dsa_df_manipulado.csv', index = False)

## Combinação de strings em DataFrames do Pandas
A função cat() pode ser usada para concatnar strings em um DataFrame do Pandas.
Vamos criar uma nova coluna concatenando ID_Pedido com Segmento como o separador "-".

In [87]:
#Concatenardo Strings
dsa_df['Pedido_Segmento'] = dsa_df['ID_Pedido'].str.cat(dsa_df['Segmento'],sep='-' )

In [88]:
dsa_df

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0       152156    16-11-08   AX-12520   Consumer  United States  South   
1       152156    16-11-08   AX-12520   Consumer  United States  South   
2       138688    16-06-12   DV-13045  Corporate  United States   West   
3       108966    15-10-11   SO-20335   Consumer  United States  South   
4       108966    15-10-11   SO-20335   Consumer  United States  South   
...        ...         ...        ...        ...            ...    ...   
9989    110422    14-01-21   TB-21400   Consumer  United States  South   
9990    121258    17-02-26   DB-13060   Consumer  United States   West   
9991    121258    17-02-26   DB-13060   Consumer  United States   West   
9992    121258    17-02-26   DB-13060   Consumer  United States   West   
9993    119914    17-05-04   CC-12220   Consumer  United States   West   

           ID_Produto        Categoria  \
0     FUR-BO-10001798        Furniture   
1     FUR-CH-10000454        Furniture   
2     OFF-LA-10000240  Office Supplies   
3     FUR-TA-10000577        Furniture   
4     OFF-ST-10000760  Office Supplies   
...               ...              ...   
9989  FUR-FU-10001889        Furniture   
9990  FUR-FU-10000747        Furniture   
9991  TEC-PH-10003645       Technology   
9992  OFF-PA-10004041  Office Supplies   
9993  OFF-AP-10002684  Office Supplies   

                                           Nome_Produto  Valor_Venda  \
0                     Bush Somerset Collection Bookcase     261.9600   
1     Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400   
2     Self-Adhesive Address Labels for Typewriters b...      14.6200   
3         Bretford CR4500 Series Slim Rectangular Table     957.5775   
4                        Eldon Fold 'N Roll Cart System      22.3680   
...                                                 ...          ...   
9989                             Ultra Door Pull Handle      25.2480   
9990  Tenex B1-RE Series Chair Mats for Low Pile Car...      91.9600   
9991                              Aastra 57i VoIP phone     258.5760   
9992  It's Hot Message Books with Stickers, 2 3/4" x 5"      29.6000   
9993  Acco 7-Outlet Masterpiece Power Center, Wihtou...     243.1600   

      Quantidade   Ano   Pedido_Segmento  
0            3.0  2016   152156-Consumer  
1            3.0  2016   152156-Consumer  
2            2.0  2016  138688-Corporate  
3            5.0  2015   108966-Consumer  
4            2.0  2015   108966-Consumer  
...          ...   ...               ...  
9989         3.0  2014   110422-Consumer  
9990         2.0  2017   121258-Consumer  
9991         2.0  2017   121258-Consumer  
9992         4.0  2017   121258-Consumer  
9993         2.0  2017   119914-Consumer  

[9994 rows x 13 columns]

## Construção de Gráficos a partir de DataSets Pandas.


In [12]:
#Instala a versão exata do Scikit-learn
!pip install -q scikit-learn==1.2.1

In [13]:
import sklearn

ModuleNotFoundError: No module named 'sklearn'

In [14]:
from sklearn.datasets import load_iris

ModuleNotFoundError: No module named 'sklearn'